### ISSUE TITLE
Schema already declared on first addSchema within decorated function

### ISSUE THREAD

**Issue by SkeLLLa** [fastify/fastify#1456]
## 🐛 Bug Report

If `.addSchema` is called within decorated function (using `fastify.decorate`) it throws an error that such schema already exists, however it is called only once

## To Reproduce

```js
const fp = require('fastify-plugin');
const app = require('fastify')();

app.register(
  fp(async (fastify) => {
    fastify.decorate('myAddSchema', function() {
      console.log('add schema called');
      fastify.addSchema({
        $id: 'sch1',
        type: 'string',
      });
    });
  })
);
app.register(async (fastify) => {
  fastify.register(async (f) => {
    f.myAddSchema({
      $id: 'sch2', // this params actually don't affect behavior
      type: 'string', // we can just call decorated function
    });
  });
});

(async () => {
  await app.ready();
  await app.listen(3000);
  console.log('app started');
})();
```

## Expected behavior

App starts without errors.

## Your Environment

- *node version*: 11.10.0
- *fastify version*: >1.13.4 & <2.0.0
- *os*: Mac, Linux

NOTE: this bug reproduces only on fastify v1.X.X. On fastify v2.X.X it works fine.

**Comment by mcollina** [comment]
That’s a new feature of Fastify v2. We decided not to backport it, because it could change the behavior of users.

V2 is going out soon, the likely solution is to upgrade.

**Comment by SkeLLLa** [comment]
@mcollina I've made some investigation. It appears that code above works fine in fastify `1.13.4`. So it seems that it was broken in `1.14`.

**Comment by mcollina** [comment]
cc @Eomm can you take a look? This is likely caused by one of your latest changes.

**Comment by SkeLLLa** [comment]
One more clarification: version 1.14.0 already has this bug, so bug is hiding somewhere between 1.13.4 and 1.14.0.

**Comment by Eomm** [comment]
The error is caused by:

https://github.com/fastify/fastify/blob/772bfbe012426d4bdffcf743859bb692656cb713/fastify.js#L752

**What is happening?**

That code produces a tree of encapsulated context

```
app
  |- register "fastify" decorated
  |- register "fastify"
      |- register "f"
```

So, when you run the `myAddSchema` function, you add the schema to "fastify decorated" and all the tree, but since the "fastify decorated" instance is the "app instance" itself, it tries to add two times the schema and it breaks.

Could you specify better the use case?

Because, to hotfix your problem, you have just to define the instance where add the schema like this:

```js
app.register(
  fp(async (fastify) => {
    fastify.decorate('myAddSchema', function (whereAddTheSchema) {
      console.log('add schema called');
      whereAddTheSchema.addSchema({
        $id: 'sch1',
        type: 'string',
      });
    });
  })
);

app.register(async (fastify) => {
  fastify.register(async (f) => {
    f.myAddSchema(f);
  });
});
```

I'm not sure if the error is that `app` has registered itself as a child or if it is needed a filter on
"some" children before executing the `addSchema` to them.

**Comment by SkeLLLa** [comment]
@Eomm there are several cases. One of them - adding global schemes that will be used in all routes, services, etc., so `whereAddTheSchema` should be "root" fastify instance.

**Comment by Eomm** [comment]
Understood, I'm checking how it could be solved.

Only as memo, the change comes from this PR https://github.com/fastify/fastify/pull/1370

**Comment by SkeLLLa** [comment]
Just as a workaround probably `addSchema` could generate some hash of schema object and compare it with existing schema or use `deepEqual` to skip adding duplicate schemas. But I think there should be more elegant solution.

**Comment by mcollina** [comment]
@Eomm one option could be to revert #1370 in v1. In fact, it was a breaking change.

**Comment by Eomm** [comment]
@mcollina I just opened a PR with the fix, could you have a look?
For sure if you think is better to revert the old PR I can arrange a PR for this.

**Comment by Eomm** [comment]
Fix released in [v1.14.2](https://github.com/fastify/fastify/releases/tag/v1.14.2)
Thank you for reporting ✌

Closing
